content/guides/server-sent-events.adoc

```diff
@@ -10,26 +10,164 @@ toc::[]
+channel of the https://clojurians.slack.com[Clojurians Slack team.]
+
 == Where We Are Going
 
+We'll be creating a new Pedestal service project, adding a simple route to stream a counter using server-sent events, and add a simple HTML page that uses a JavaScript EventSource to open a connection to the stream.
+
 == Before We Begin
 
-== Content Section 1
+This guide will use http://leiningen.org/[Leiningen] to generate and
+build a template project. If you haven't already installed it, please
+take a few minutes to set it up.
+
+== Generate Our New Project
+
+----
+$ lein new pedestal-service myapp
+Generating a pedestal-service application called server-sent-events.
+$ cd server-sent-events
+----
+
+== Add Hiccup
+
+We will be using Hiccup to generate our index page, so we will need to add the dependency in our project.clj.
+
+[source,clojure]
+.project.clj
+----
+include::server-sent-events/project.clj[tags=add-hiccup]
+----
+
+== Clean Up
+
+Here's the service.clj file generated for us:
+
+[source,clojure]
+.src/server_sent_events/service.clj
+----
+include::server-sent-events/src/server_sent_events/service.clj[tags=original-service]
+----
+
+Let's simply things by uncommenting the 'Terse/Vector-based routes' section and removing uneccessary code.
+
+[source,clojure]
+.src/server_sent_events/service.clj
+----
+include::server-sent-events/src/server_sent_events/service.clj[tags=cleaned-up]
+----
+
+== Server-Sent Event Route
+
+Now let's add our server-sent event route. First Lets require SSE and core.async.
+
+[source,clojure]
+.src/server_sent_events/service.clj
+----
+include::server-sent-events/src/server_sent_events/service.clj[tags=require-sse]
+----
+
+Now we'll add our counter route. We create the "/counter" route by calling sse/start-event-stream and passing in our stream-ready handler. This creates an interceptor that will handle setting up the connection and call the stream-ready handler. A stream-ready handler receives a core.async event-chan and SSE context as parameters. You send events on the channel using regular core.async (such as async/>!!) functions and terminate the connection by calling async/close! on the channel. Here our event consists the :name and :data keys, but could also contain an :id for the event.
+
+[source,clojure]
+.src/server_sent_events/service.clj
+----
+include::server-sent-events/src/server_sent_events/service.clj[tags=counter-route]
+----
+
+Now lets start up our server.
+
+----
+$ lein run
+...
+Creating your server...
+...
+INFO  org.eclipse.jetty.server.Server - Started @4434ms
+----
+
+And checkout out our event stream in another terminal using curl.
+
+----
+$ curl http://localhost:8080/counter
+event: counter
+data: 0
+
+event: counter
+data: 1
+
+event: counter
+data: 2
+
+event: counter
+data: 3
+
+event: counter
+data: 4
+
+event: counter
+data: 5
+
+event: counter
+data: 6
+
+event: counter
+data: 7
+
+event: counter
+data: 8
+
+event: counter
+data: 9
```

Comment:
It's worth noting that the data will always be received as a string. If you want to send JSON, it's up to the sender to encode it and the receiver to decode it.

Also, it's important to note that none of the Pedestal interceptors are invoked when sending SSE. (This has caused confusion for some users.) The interceptors are used for the initial connection request from the client, but not on the events themselves.